File: components/link.rb
require 'components/attachment'
require 'mathcontext'

module Cpg::Components
	class Link < Attachment
		property :from, :to, :act_on, :equation, :label
		read_only :from, :to
	
		def self.new(to = nil)
			# generate a new Link object for each pair of first -> next
			return super([nil, nil]) if to == nil
		
			res = []
		
			# self connection
			if to.length == 1
				res << super([to.first, to.first])
			else		
				to[1..-1].each do |other|
					res << super([to.first, other])
				end
			end

			res.length > 1 ? res : res.first
		end
	
		def set_property(name, val)
			super
			
			if name == 'from'
				@objects[0] = val
			elsif name == 'to'
				@objects[1].unlink(self) if @objects[1] != nil
				@objects[1] = val
				@objects[1].link(self) if val
			elsif name == 'label' || name == 'equation'
				request_redraw
			end
			
			true
		end
	
		def initialize(to = [nil, nil])
			super(to)
			
			@offset = 0
		
			if to.length > 1
				self.from = @objects[0]
				self.to = @objects[1]
			
				self.to.link(self) if self.to != nil
			end
		end
		
		def <<(objs)
			objs = objs.is_a?(Enumerable) ? objs : [objs]
			
			if objs.length == 1 && !self.from && !self.to
				objs << objs[0]
			end

			if self.from == nil
				self.from = objs.shift
				@objects[0] = self.from
			end
			
			if self.to == nil && !objs.empty?
				self.to = objs.shift
				self.to.link(self)

				@objects[1] = self.to
			end
		end
	
		def removed
			self.to.unlink(self) if self.to != nil
		end
	
		def calc_control(from, to)
			dx = to[0] - from[0]
			dy = to[1] - from[1]
		
			x = from[0] + dx / 2.0
			y = from[1] + dy / 2.0
		
			same = (dy == 0 && dx == 0)
		
			# and offset it perpendicular
			dist = 1 * (@offset.to_f + 1)
		
			alpha = same ? 0 : Math.atan(dx.to_f / -dy)
			alpha += Math::PI if dy >= 0
		
			dy = Math.sin(alpha) * dist
			dx = Math.cos(alpha) * dist
		
			return x + dx, y + dy
		end
	
		def draw(ct)
			# draw arrow from o1 to o2
			return if @objects.empty? or @objects.first == nil
		
			a1 = @objects[0].allocation
			a2 = @objects[1].allocation
		
			from = [mean(a1, 0), mean(a1, 1)]
			to = [mean(a2, 0), mean(a2, 1)]
					
			ct.move_to(from[0], from[1])
		
			x, y = calc_control(from, to)

			if from[0] == to[0] and from[1] == to[1]
				# draw pretty one
				yy = (@offset + 1) + 0.5
				xx = 2
				ct.curve_to(to[0] - xx, to[1] - yy, to[0] + xx, to[1] - yy, to[0], to[1])
			else
				ct.curve_to(x, y, x, y, to[0], to[1])
			end
		
			if @selected	
				ct.set_source_rgb(0.8, 0.8, 1)
			else
				ct.set_source_rgb(0.8, 0.8, 0.8)
			end
			
			if @focus
				ct.set_dash([ct.line_width * 5, ct.line_width * 5], 0)
			end
		
			ct.line_width = ct.line_width * 2
			ct.stroke
		
			size = 0.15
		
			# draw the arrow, first move to the center, then rotate, then draw
			# the arrow
			if from[0] == to[0] and from[1] == to[1]
				x = evaluate_bezier(to[0], to[0] - xx, to[0] + xx, to[0], 0.5)
				y = evaluate_bezier(to[1], to[1] - yy, to[1] - yy, to[1], 0.5)
			
				pos = -0.5 * Math::PI
			else
				x = evaluate_bezier(from[0], x, x, to[0], 0.5)
				y = evaluate_bezier(from[1], y, y, to[1], 0.5)

				xx = to[0] - from[0]
				yy = to[1] - from[1]
		
				if xx == 0
					pos = yy < 0 ? 1.5 * Math::PI : Math::PI * 0.5
				else
					pos = Math::atan(yy.to_f / xx)
		
					pos += Math::PI if xx < 0
					pos += Math::PI * 2 if xx > 0 && yy < 0
				end
		
				pos += Math::PI * 0.5
			end
		
			ct.translate(x, y)		
			ct.move_to(0, 0)
			ct.rotate(pos)
		
			ct.rel_line_to(-size, 0)
			ct.rel_line_to(size, -size)
			ct.rel_line_to(size, size)
			ct.rel_line_to(-size, 0)
		
			ct.fill
		
			# also draw some text here
			return unless (self.equation or self.label)
			t = to_s

			e = ct.text_extents(t)
			ct.rotate(0.5 * Math::PI)
		
			ct.rotate(Math::PI) if from[0] < to[0]
		
			ct.move_to(-e.width / 2.0, -e.height * 1.5)
			ct.set_source_rgb(0, 0, 0)
			ct.show_text(t)
		end
	
		def to_s
			(self.label or self.equation).to_s
		end
	
		def mean(alloc, offset)
			return alloc[offset] + alloc[offset + 2] / 2.0
		end
	
		def distance_to_line(start, stop, point)
			x, y = *point
		
			dx = [x - start[0], stop[0] - start[0]]
			dy = [y - start[1], stop[1] - start[1]]
		
			dot = dx[0] * dx[1] + dy[0] * dy[1]
			len_sq = dx[1] * dx[1] + dy[1] * dy[1]
			param = dot / len_sq

			if param < 0
				xx = start[0]
				yy = start[1]
			elsif param > 1
				xx = stop[0]
				yy = stop[1]
			else
				xx = start[0] + param * dx[1]
				yy = start[1] + param * dy[1]
			end

			Math.sqrt((x - xx) * (x - xx) + (y - yy) * (y - yy))
		end
	
		def evaluate_bezier(p0, p1, p2, p3, t)
			((1 - t)**3) * p0 + 3 * t * ((1 - t)**2) * p1 + 3 * (t**2) * (1 - t) * p2 + (t**3) * p3
		end
	
		def rect_hittest(p1, p2, p3, p4, rect, factor)
			# different hittest strategy, take points on the curve and see if
			# they are withing the rect
			rect = rect.collect { |x| x / factor.to_f }
		
			num = 5
			(0..num).each do |t| 
				px = evaluate_bezier(p1[0], p2[0], p3[0], p4[0], t / num.to_f)
				py = evaluate_bezier(p1[1], p2[1], p3[1], p4[1], t / num.to_f)
			
				return true if rect[0] <= px && rect[2] >= px && rect[1] <= py && rect[3] >= py
			end
		
			false
		end
	
		def hittest(rect, factor)
			dx = (rect[2] - rect[0]).abs
			dy = (rect[3] - rect[1]).abs
	
			a1 = @objects[0].allocation
			a2 = @objects[1].allocation
		
			# find P0 and P3 for bezier curve
			from = [mean(a1, 0), mean(a1, 1)]
			to = [mean(a2, 0), mean(a2, 1)]
		
			# find control point P1 and P2 for bezier curve
			cx, cy = calc_control(from, to)
		
			# do piece wise linearization
			num = 5
			dist = []
		
			prevp = from
		
			if from[0] == to[0] && from[1] == to[1]
				p2 = [to[0] - 2, to[1] - ((@offset + 1) + 0.5)]
				p3 = [to[0] + 2, to[1] - ((@offset + 1) + 0.5)]
			else
				p2 = [cx, cy]
				p3 = [cx, cy]
			end
		
			if dx > 1 || dy > 1
				return rect_hittest(from, p2, p3, to, rect, factor)
			end
		
			(1..num).each do |i|
				px = evaluate_bezier(from[0], p2[0], p3[0], to[0], i / num.to_f)
				py = evaluate_bezier(from[1], p2[1], p3[1], to[1], i / num.to_f)

				l = [prevp[0], prevp[1], px, py]

				dist << distance_to_line(prevp, [px, py], [rect[0] / factor.to_f, rect[1] / factor.to_f])
				prevp = [px, py]
			end
		
			dist.min < 10 / factor.to_f
		end
	
		def self.limits
			[1, -1]
		end
	
		def simulation_reset
		end
	
		def to_xml
			super
		end
	
		def simulation_evaluate
			begin
				return {} if self.act_on == nil || self.act_on == '' || self.equation == nil || self.equation == ''
			rescue StandardError, e
				p e
			end

			# create execution context
			c = Cpg::MathContext.new(Cpg::Simulation.instance.state, self.from.state, state, {'from' => self.from, 'to' => self.to})
		
			vars = self.act_on.to_s.split(/\s*,\s*/)
			eq = self.equation.to_s.split(/\s*,\s*/, vars.length)
		
			if vars.length != eq.length
				STDERR.puts('Number of variables is not number of equations')
				return {}
			end
		
			res = {}
		
			vars.each_with_index do |on, i|
				res[on] = c.eval(eq[i])
			end
	
			return res
		end
	end
end

File: components/Link.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace Cpg.Studio.Components
{
	public class Link : Simulated
	{
		public class Action
		{
			Link d_link;
			LinkAction d_action;
			
			public Action(Link link, LinkAction action)
			{
				d_link = link;
				d_action = action;
			}
			
			public Action() : this(null, null)
			{
			}

			public string Target
			{
				get
				{
					return d_action.Target.Name;
				}
				set
				{
					d_action = d_link.UpdateAction(this, value, Equation);
				}
			}
			
			public Cpg.Property Property
			{
				get
				{
					return d_action.Target;
				}
			}
			
			public LinkAction LinkAction
			{
				get
				{
					return d_action;
				}
			}

			public string Equation
			{
				get
				{
					return d_action.Expression.AsString;
				}
				set
				{
					d_action = d_link.UpdateAction(this, d_action.Target.Name, value);
				}
			}
		}
		
		Components.Simulated d_from;
		Components.Simulated d_to;
		
		int d_offset;
		
		double[] d_selectedColor;
		double[] d_normalColor;
		double[] d_hoverColor;
		
		public Link(Cpg.Link obj, Components.Simulated from, Components.Simulated to) : base(obj)
		{
			if (from == null && obj != null)
				From = Components.Simulated.FromCpg(obj.From);
			else
				From = from;
			
			if (to == null && obj != null)
				To = Components.Simulated.FromCpg(obj.To);
			else
				To = to;
			
			d_normalColor = new double[] {0.7, 0.7, 0.7, 0.6};
			d_selectedColor = new double[] {0.6, 0.6, 1, 0.6};
			d_hoverColor = new double[] {0.3, 0.6, 0.3, 0.6};
		}
		
		public Link(Cpg.Link obj) : this(obj, null, null)
		{
		}
		
		public Link() : this(null, null, null)
		{
		}
		
		public Components.Object[] Objects
		{
			get
			{
				if (d_object != null)
					return new Components.Object[] {d_from, d_to};
				else
					return new Components.Object[] {};
			}
		}
		
		public int Offset
		{
			get
			{
				return d_offset;
			}
			set
			{
				d_offset = value;
			}
		}
		
		public bool Empty()
		{
			return d_object == null;
		}
		
		public bool SameObjects(Components.Link other)
		{
			return (other.d_from == d_from && other.d_to == d_to);
		}
		
		public Components.Simulated From
		{
			get
			{
				return d_from;
			}
			set
			{
				d_from = value;
				
				if (d_from != null)
				{
					d_from.Moved += delegate(object sender, EventArgs e) {
						DoRequestRedraw();
					};
				}
			}
		}
		
		public Components.Simulated To
		{
			get
			{
				return d_to;
			}
			set
			{
				if (d_to != null)
					d_to.Unlink(this);
					
				d_to = value;
				
				if (d_to != null)
				{
					d_to.Link(this);
				}
			}
		}
		
		public Action[] Actions
		{
			get
			{
				List<Action> actions = new List<Action>();
				
				foreach (LinkAction action in (d_object as Cpg.Link).Actions)
				{
					actions.Add(new Action(this, action));
				}
				
				return actions.ToArray();
			}
		}
		
		public Action AddAction(string property, string expression)
		{
			Cpg.Property prop = d_to.Object.Property(property);
			
			if (prop != null)
				return new Action(this, (d_object as Cpg.Link).AddAction(prop, expression));
			else
				return null;
		}
		
		public bool RemoveAction(Action action)
		{
			return (d_object as Cpg.Link).RemoveAction(action.LinkAction);
		}
		
		public Cpg.LinkAction UpdateAction(Action action, string property, string expression)
		{
			Cpg.Link link = d_object as Cpg.Link;
			Cpg.Property prop = link.To.Property(property); 

			if (prop == null)
				return action.LinkAction;
				
			if (!RemoveAction(action))
				return action.LinkAction;

			return link.AddAction(prop, expression);
		}
		
		public override void Removed()
		{
			if (d_to != null)
				d_to.Unlink(this);

			base.Removed();
		}

		private bool RectHittest(PointF p1, PointF p2, PointF p3, PointF p4, Allocation rect, float gridSize)
		{
			Allocation other = new Allocation(0, 0, 1f / gridSize, 1f / gridSize);
			
			for (int i = 0; i < 5; ++i)
			{
				other.X = EvaluateBezier(p1.X, p2.X, p3.X, p4.X, (float)i / 5);
				other.Y = EvaluateBezier(p1.Y, p2.Y, p3.Y, p4.Y, (float)i / 5);
				
				if (rect.Intersects(other))
					return true;
			}
			
			return false;
		}
		
		private float DistanceToLine(PointF start, PointF stop, PointF point)
		{
			PointF dx = new PointF(point.X - start.X, stop.X - start.X);
			PointF dy = new PointF(point.Y - start.Y, stop.Y - start.Y);
			
			float dot = dx.X * dx.Y + dy.X * dy.Y;
			float len_sq = dx.Y * dx.Y + dy.Y * dy.Y;
			float param = dot / len_sq;
			
			PointF res = new PointF(0, 0);
			
			if (param < 0)
			{
				res = start;
			}
			else if (param > 1)
			{
				res = stop;
			}
			else
			{
				res.X = start.X + param * dx.Y;
				res.Y = start.Y + param * dy.Y;
			}
			
			return (float)Math.Sqrt((point.X - res.X) * (point.X - res.X) + (point.Y - res.Y) * (point.Y - res.Y));
		}
		
		public bool HitTest(Allocation rect, int gridSize)
		{			
			PointF[] points = ControlPoints();
			
			// Piece wise linearization
			int num = 5;
			List<float> dist = new List<float>();
			
			if (rect.Width * gridSize > 1.5 || rect.Height * gridSize > 1.5)
			{
				return RectHittest(points[0], points[1], points[2], points[3], rect, gridSize);
			}

			PointF prevp = points[0];

			for (int i = 1; i <= num; ++i)
			{
				PointF pt = EvaluateBezier(points[0], points[1], points[2], points[3], (float)i / num);
				dist.Add(DistanceToLine(prevp, pt, new PointF(rect.X, rect.Y)));
				
				prevp = pt;
			}

			return (Utils.Min(dist) < 10.0f / gridSize);
		}
		
		
		private double[] StateColor()
		{
			if (Selected)
			{
				return d_selectedColor;
			}
			else if (MouseFocus)
			{
				return d_hoverColor;
			}
			else
			{
				return d_normalColor;
			}
		}
		
		private PointF CalculateControl(PointF from, PointF to)
		{
			PointF diff = new PointF(to.X - from.X, to.Y - from.Y);
			PointF point = new PointF(from.X + diff.X / 2, from.Y + diff.Y / 2);
			
			bool same = (diff.X == 0 && diff.Y == 0);
			
			// Offset perpendicular
			float dist = 1 * (d_offset + 1);
			float alpha = same ? 0 : (float)Math.Atan(diff.X / -diff.Y);
			
			if (diff.Y >= 0)
				alpha += (float)Math.PI;
		
			return new PointF(point.X + (float)Math.Cos(alpha) * dist, point.Y + (float)Math.Sin(alpha) * dist);
		}
		
		public float EvaluateBezier(float p0, float p1, float p2, float p3, float t)
		{
			return (float)Math.Pow(1 - t, 3) * p0 + 3 * t * (float)Math.Pow(1 - t, 2) * p1 + 3 * (float)Math.Pow(t, 2) * (1 - t) * p2 + (float)Math.Pow(t, 3) * p3;
		}
		
		public PointF EvaluateBezier(PointF p0, PointF p1, PointF p2, PointF p3, float t)
		{
			return new PointF(
				EvaluateBezier(p0.X, p1.X, p2.X, p3.X, t),
				EvaluateBezier(p0.Y, p1.Y, p2.Y, p3.Y, t)
			);
		}
		
		private PointF[] ControlPoints()
		{
			if (d_from == null || d_to == null)
				return null;

			Allocation a1 = d_from.Allocation;
			Allocation a2 = d_to.Allocation;
			
			PointF from = new PointF(a1.X + a1.Width / 2, a1.Y + a1.Height / 2);
			PointF to = new PointF(a2.X + a2.Width / 2, a2.Y + a2.Height / 2);
			PointF control = CalculateControl(from, to);
			
			if (from.X == to.X && from.Y == to.Y)
			{
				PointF pts = new PointF(2, (d_offset + 1) + 0.5f);

				return new PointF[] {
					from,
					new PointF(to.X - pts.X, to.Y - pts.Y),
					new PointF(to.X + pts.X, to.Y - pts.Y),
					to
				};
			}
			else
			{
				return new PointF[] {from, control, control, to};
			}
		}
		
		public override void Draw(Cairo.Context graphics)
		{			
			PointF[] points = ControlPoints();
			
			if (points == null)
				return;
			
			graphics.Save();			

			double[] color = StateColor();
			double uw = graphics.LineWidth;
			
			if (KeyFocus)
			{
				graphics.LineWidth *= 4;
				graphics.SetDash(new double[] {graphics.LineWidth, graphics.LineWidth}, 0);
			}
			else if (MouseFocus)
			{
				graphics.LineWidth *= 2;
			}

			graphics.MoveTo(points[0].X, points[0].Y);
			graphics.CurveTo(points[1].X, points[1].Y, points[2].X, points[2].Y, points[3].X, points[3].Y);
			graphics.SetSourceRGBA(color[0], color[1], color[2], color[3]);

			graphics.Stroke();
			
			PointF xy;
			float pos;
			
			float size = 0.15f;

			// Draw the arrow, first move to the center, then rotate, then draw the arrow
			xy = EvaluateBezier(points[0], points[1], points[2], points[3], 0.5f);

			if (points[0] == points[3])
			{
				pos = 1.5f * (float)Math.PI;
			}
			else
			{
				PointF diff = new PointF(points[3].X - points[0].X, points[3].Y - points[0].Y);
				
				if (diff.X == 0)
				{
					pos = (diff.Y < 0 ? 1.5f : 0.5f) * (float)Math.PI; 
				}
				else
				{
					pos = (float)Math.Atan(diff.Y / diff.X);
					
					if (diff.X < 0)
						pos += (float)Math.PI;
					else if (diff.Y < 0)
						pos += 2 * (float)Math.PI;
				}
				
				pos += 0.5f * (float)Math.PI;
			}
			
			graphics.Translate(xy.X, xy.Y);
			graphics.MoveTo((pos <= Math.PI ? -1 : 1) * size / 2, 0);
			graphics.Rotate(pos);

			graphics.RelLineTo(-size, 0);
			graphics.RelLineTo(size, -size);
			graphics.RelLineTo(size, size);
			graphics.RelLineTo(-size, 0);
			
			graphics.Fill();
			
			string s = ToString();
			
			if (s == String.Empty)
			{
				graphics.Restore();
				return;
			}
		
			graphics.Rotate(0.5 * Math.PI);
			
			if (points[0].X < points[3].X)
				graphics.Rotate(Math.PI);
		
			graphics.Scale(uw, uw);
			Pango.Layout layout = Pango.CairoHelper.CreateLayout(graphics);
			
			layout.FontDescription = Settings.Font;
			layout.SetText(s);
			
			if (MouseFocus)
				graphics.SetSourceRGB(d_hoverColor[0], d_hoverColor[1], d_hoverColor[2]);
			else
				graphics.SetSourceRGB(0.3, 0.3, 0.3);
			
			int width, height;
			
			layout.GetSize(out width, out height);
			width = (int)(width / Pango.Scale.PangoScale);
			height = (int)(height / Pango.Scale.PangoScale);
			
			graphics.MoveTo(-width / 2.0, -height * 1.5);
			
			Pango.CairoHelper.ShowLayout(graphics, layout);

			graphics.Restore();
		}

		public override Allocation Extents(float scale, Cairo.Context graphics)
		{
			PointF[] points = ControlPoints();
			
			if (points == null)
				return new Allocation(0, 0, 0, 0);
			
			List<float> xx = new List<float>();
			List<float> yy = new List<float>();
			
			for (int i = 0; i < points.Length; ++i)
			{
				xx.Add(points[i].X * scale);
				yy.Add(points[i].Y * scale);
			}
			
			float minx = Utils.Min(xx);
			float maxx = Utils.Max(xx);
			float miny = Utils.Min(yy);
			float maxy = Utils.Max(yy);
			
			int width, height;
			MeasureString(graphics, ToString(), out width, out height);

			return new Allocation(minx - height, miny - height, maxx - minx + height * 2, maxy - miny + height * 2);
		}
		
		public override bool CanIntegrate
		{
			get
			{
				return false;
			}
		}
	}
}
